pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -363,12 +363,69 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueue(t *testing.T) {
 	q.Add(&medPriorityPod)
 	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(&unschedulablePod), q.SchedulingCycle())
 	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(&highPriorityPod), q.SchedulingCycle())
+	// sleep 1 second so that pods is guaranteed to be added into unschedulableQ first and then
+	// the move request is happened later, otherwise there is potential a test flake here, as the
+	// move request has a chance to be finished before the pods get into the unschedulableQ.
+	// In that case, the pods will be moved to backoffQ even `QueuedPodInfo.Attempts` is zero.
+	time.Sleep(1000 * time.Millisecond)
 	q.MoveAllToActiveOrBackoffQueue("test")
+	if q.activeQ.Len() != 3 {
+		t.Error("Expected 3 item to be in activeQ")
+	}
+	if q.podBackoffQ.Len() != 0 {
+		t.Error("Expected 0 items to be in podBackoffQ")
+	}
+}
+
+// TestPriorityQueue_RaceConditionOnScheduling mimic a race condition could be found with preemption.
+// In this particular case, the high priority pod is stuck in the backoffQ and if new pod is coming,
+// regardless it's priority, the new pod will be scheduled first.
+// The preemption could happen again but there is a chance the pod will never got scheduled as long as
+// there is new pod coming for scheduling since the high priority pod could be moved from backoffQ
+// to activeQ only when the time for backoff expiration is up.
+// see https://github.com/kubernetes/kubernetes/issues/93505 for the details.
+func TestPriorityQueue_RaceConditionOnScheduling(t *testing.T) {
+	q := NewPriorityQueue(newDefaultQueueSort())
+	// comes a lowPriority pod and is added to activeQ, this mimic a pod that is managed by
+	// replication controller is preempted by the `highPriorityPod` and a new pod is going to
+	// be re-scheduled.
+	q.Add(&unschedulablePod)
+	// events that make the move request is happen, pod deletion, service add etc.
+	// which will trigger the move request.
+	q.MoveAllToActiveOrBackoffQueue("test")
+	// sleep 1 second in order to guarantee the move request is happened before the highPriorityPod
+	// wants to be added to unschedulableQ.
+	time.Sleep(1000 * time.Millisecond)
+	// even the unschedulableQ is empty and `QueuedPodInfo.Attempts` is empty, the highPriorityPod
+	// is pushed to backoffQ instead of unschedulableQ.
+	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(&highPriorityPod), q.SchedulingCycle())
 	if q.activeQ.Len() != 1 {
 		t.Error("Expected 1 item to be in activeQ")
 	}
-	if q.podBackoffQ.Len() != 2 {
-		t.Error("Expected 2 items to be in podBackoffQ")
+	if q.podBackoffQ.Len() != 1 {
+		t.Error("Expected 1 items to be in podBackoffQ")
+	}
+	// the lowPriority pod is popped and got scheduled while the highPriorityPod is stuck in the
```

Comment:
This is by design to avoid starvation from high priority pods. Let's discuss in the original issue. However, we intend to reduce the problems of this with #94009

Reply:
problem here is that the `highPriorityPod` should not be in `backoffQ` in the first place, it is moved to be `backoffQ` just because there was a move request happened first.
`lowPriority` pod popped here should be fine, but there should a way for the preemption to get convergent,  pls see the code following, 
```golang
	// another pod is added to activeQ.
	q.Add(&unschedulablePod)
	if q.activeQ.Len() != 1 {
		t.Error("Expected 1 item to be in activeQ")
	}
```
it's possible a new pod is added to Queue while the `highPriorityPod` is still backoff-ing, after the backoff is up, the `highPriorityPod` needs to preempt the `lowPriority` pod again.

Reply:
I prefer this to be split in a different PR so we can already merge the tests using the deterministic clock.